### PR TITLE
Sharding propagation pass

### PR DESCRIPTION
**Pull request by cowanmeg**
A simple sharding propagation pass that shards TensorViews without a DeviceMesh like it's first producer TensorView with a DeviceMesh.
Assumes global TensorViews are manually sharded.

### CODE REVIEW

File: tests/cpp/test_multidevice_sharding.cpp [hunk 1]
```diff
@@ -40,6 +40,30 @@ TEST_F(NVFuserTest, TestIsSharded) {
   EXPECT_ANY_THROW(isSharded(c));
 }
 
+TEST_F(NVFuserTest, TestPropagateSharding) {
```
Comment (wujingyue) [inline tests/cpp/test_multidevice_sharding.cpp:43]:
```suggestion
TEST_F(ShardingTest, PropagateSharding) {
```
and at the beginning of the file say `using ShardingTest = NVFuserTest`. I'd do this for other NVFuserTests in this file as well. This gives the tests a less overloaded fixture name for easy filtering. For example, `bin/nvfuser_tests --gtest_filter=ShardingTest*`.

File: tests/cpp/test_multidevice_sharding.cpp [hunk 2]
```diff
@@ -40,6 +40,30 @@ TEST_F(NVFuserTest, TestIsSharded) {
...
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
```
Comment (wujingyue) [inline tests/cpp/test_multidevice_sharding.cpp:44]:
```suggestion
  Fusion fusion;
```
I don't see a need to use a smart pointer.

**Review by wujingyue — COMMENTED** [review]
Thanks for splitting up the PR!

File: tests/cpp/test_multidevice_sharding.cpp [hunk 3]
```diff
@@ -40,6 +40,30 @@ TEST_F(NVFuserTest, TestIsSharded) {
...
+  FusionGuard fg(fusion.get());
+
+  TensorView* a = makeSymbolicTensor(3);
+  TensorView* b = makeSymbolicTensor(3);
+  TensorView* c = add(a, b);
```
Comment (samnordmann) [inline tests/cpp/test_multidevice_sharding.cpp:49]: just for my own understanding: does this `add` causes the automatic insertion of (one or two?) castOps? proably some bcast too right?

File: csrc/multidevice/utils.cpp [hunk 4]
```diff
@@ -225,6 +225,30 @@ void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
 }
 } // namespace
 
+void propagateShardings(Fusion* fusion) {
+  for (auto expr : fusion->exprs()) {
+    auto inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+    auto outputs = ir_utils::filterByType<TensorView>(expr->outputs());
+    std::vector<TensorView*> input_with_mesh;
```
Comment (samnordmann) [inline csrc/multidevice/utils.cpp:232]: it is enough to simply allocate one pointer `TensorView*`, initialized to `nullptr`

File: csrc/multidevice/utils.cpp [hunk 5]
```diff
@@ -225,6 +225,30 @@ void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
...
+    for (auto tv : inputs) {
+      if (tv->hasDeviceMesh()) {
+        input_with_mesh.push_back(tv);
+      }
```
Comment (samnordmann) [inline csrc/multidevice/utils.cpp:236]: `break`

**Review by samnordmann — COMMENTED** [review]
Looks good, thanks!! I only left very minor comments

File: csrc/multidevice/utils.cpp [hunk 6]
```diff
@@ -225,6 +225,30 @@ void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
...
+        !input_with_mesh.empty(),
+        "At least one input requires a DeviceMesh ",
+        expr->toString());
+    std::vector<TensorView*> outputs_without_mesh;
+    for (auto tv : outputs) {
+      if (!tv->hasDeviceMesh()) {
+        outputs_without_mesh.push_back(tv);
+      }
+    }
+    shardAllLike(input_with_mesh[0], outputs_without_mesh);
```
Comment (wujingyue) [inline csrc/multidevice/utils.cpp:248]: Will two TensorViews in a fusion have different device mesh? I understand they can be sharded differently but I failed to see a value for them having different device meshes.

Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:248]: yes! pipeline parallel!

File: tests/cpp/test_multidevice_sharding.cpp (hunk 3 above)
Comment (cowanmeg) [reply tests/cpp/test_multidevice_sharding.cpp:49]: This add doesn't, but I purposefully didn't apply any sharding on C for this test.
If a was a half data type, and c float then an CastOp casting a to float would be automatically inserted.

File: csrc/multidevice/utils.cpp (hunk 6 above)
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:248]: A really simple example might be something like this where we have one add computed on device 0 and then the second add computed on device 1
```
TensorView* a = makeTensor(2);
TensoView* b = add(a, a);
TensorView* c = set(b);
TensorView* d = add(c, c);

a->setDeviceMesh({0});
b->setDeviceMesh({0});

c->setDeviceMesh({1});
d->setDeviceMesh({1});
```

File: tests/cpp/test_multidevice_sharding.cpp (hunk 1 above)
Comment (cowanmeg) [reply tests/cpp/test_multidevice_sharding.cpp]: That's nice way to organize our tests better. Made it MultiDeviceUtilsTest and bundled it with others :D

**Review by wujingyue — APPROVED** [review]

**Comment by cowanmeg** [comment]
!build --dist

File: tests/cpp/test_multidevice_sharding.cpp [hunk 7]
```diff
@@ -18,11 +18,13 @@
...
 namespace nvfuser {
 
+using MultiDeviceUtilsTest = NVFuserTest;
+
 // TODO: This test checks that isSharded generates an error when a split/merged
 // axis is parallelized with DIDx. Update when this restriction is lifted.
-TEST_F(NVFuserTest, TestIsSharded) {
-  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
-  FusionGuard fg(fusion.get());
+TEST_F(MultiDeviceUtilsTest, TestIsSharded) {
```
[the remark this answers is not in the document]
Comment (cowanmeg) [reply tests/cpp/test_multidevice_sharding.cpp:25]: Yeah, I think there is probably another multidevice test refactor coming up where we pull out tests that don't need multidevice execution into another file

**Comment by cowanmeg** [comment]
!build --dist

**Pull request merged**